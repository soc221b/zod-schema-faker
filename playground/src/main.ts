import { z } from 'zod'
import { fake, install } from 'zod-schema-faker'

const main = async () => {
  const schema = z.string().uuid()
  let data: string

  if (process.env.NODE_ENV === 'development') {
    install()

    data = fake(schema)
  } else {
    data = await fetch('https://httpbin.org/uuid')
      .then(response => response.json())
      .then(json => json.uuid)
  }

  const app = document.querySelector<HTMLDivElement>('#app')!
  app.innerHTML = `<pre>${JSON.stringify(schema.safeParse(data), null, 2)}</pre>`
}

main()
