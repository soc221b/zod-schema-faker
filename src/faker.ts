import { Faker, faker as _faker } from '@faker-js/faker'
import * as RandExp from 'randexp'

let _seedValue: number | undefined

// https://github.com/faker-js/faker/issues/448
// TODO: create standalone faker instead of use following workaround
export const runFake = <Runner extends (faker: Faker) => any>(
  runner: Awaited<ReturnType<Runner>> extends ReturnType<Runner> ? Runner : never,
): ReturnType<Runner> => {
  const oldSeedValue = _faker.seed()
  _faker.seed(_seedValue)

  const result = runner(_faker)
  if (result instanceof Promise) {
    throw new SyntaxError('InternalError: runFake cannot be used with async functions')
  }

  _faker.seed(oldSeedValue)

  return result
}

export const randexp = (pattern: string | RegExp, flags?: string): string => {
  const randexp = new RandExp(pattern, flags)
  randexp.randInt = (from, to) => runFake(faker => faker.number.int({ min: from, max: to }))
  return randexp.gen()
}

/**
 * sets the seed to use.
 */
export const seed = (value?: number): void => {
  _seedValue = value
}
