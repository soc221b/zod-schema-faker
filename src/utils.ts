import { z } from 'zod'
import { ZodSchemaFakerError } from './error'

export function assertsZodSchema(schema: unknown): asserts schema is z.ZodTypeAny {
  if (
    typeof schema !== 'object' ||
    schema === null ||
    '_parse' in schema === false ||
    typeof schema._parse !== 'function'
  ) {
    throw new ZodSchemaFakerError(`Expected a zod schema, but got ${schema}`)
  }
}
