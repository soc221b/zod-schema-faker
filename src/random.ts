import { Faker, fakerEN, SimpleFaker } from '@faker-js/faker'
import RandExp from 'randexp'

let faker: Faker = fakerEN

/**
 * Use given faker instance instead of the default one.
 *
 * @default fakerEN
 *
 * @see https://fakerjs.dev/guide/localization.html for more information.
 */
export function installFaker(fakerInstance: Faker): void {
  faker = fakerInstance
}

// https://github.com/faker-js/faker/issues/448
// TODO: create standalone faker instead of use following workaround
export const runFake = <Runner extends (faker: Faker) => any>(
  runner: Awaited<ReturnType<Runner>> extends ReturnType<Runner> ? Runner : never,
): ReturnType<Runner> => {
  const result = runner(faker)
  if (result instanceof Promise) {
    throw new SyntaxError('InternalError: runFake cannot be used with async functions')
  }

  return result
}

const simpleFaker = new SimpleFaker()
/**
 * Create random strings that match a given regular expression.
 */
export const randexp = (pattern: string | RegExp, flags?: string): string => {
  const randexp = new RandExp(pattern, flags)
  randexp.randInt = (from, to) => simpleFaker.number.int({ min: from, max: to })
  return randexp.gen()
}

/**
 * Sets the seed or generates a new one.
 *
 * This method is intended to allow for consistent values in tests, so you might want to use hardcoded values as the seed.
 */
export const seed = (value?: number): void => {
  faker.seed(value)
  simpleFaker.seed(value)
}
