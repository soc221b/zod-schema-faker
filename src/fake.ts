import { z } from 'zod'
import { ZodSchemaFakerError } from './error'
import { assertsZodSchema } from './utils'

export const zodTypeKindToZodTypeFaker: Map<z.ZodFirstPartyTypeKind, any /* TODO: should not use any */> = new Map()
export const zodTypeToZodTypeFaker: Map<z.ZodType, any /* TODO: should not use any */> = new Map()

/**
 * generate fake data based on schema
 */
export const fake = <T extends z.ZodType>(schema: T): z.infer<T> => {
  assertsZodSchema(schema)

  const typeName = (schema._def as any).typeName
  const faker = zodTypeToZodTypeFaker.get(schema) ?? zodTypeKindToZodTypeFaker.get(typeName)
  if (faker === undefined) {
    throw new ZodSchemaFakerError(`Unsupported schema type: ${typeName}.
- If this is a custom schema, you may not have installed a corresponding faker. If you need help with installation, please refer to the documentation for more information.
- If this is a built-in schema, it may not be implemented yet. Please file an issue to let us know: https://github.com/soc221b/zod-schema-faker/issues`)
  }

  return new faker(schema).fake()
}
