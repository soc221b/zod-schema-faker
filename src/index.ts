import { z } from 'zod'
import { fake as _fake } from './fake'
import { resetSeed } from './random'

export const fake = <T extends z.ZodType>(schema: T): z.infer<T> => {
  const result = _fake(schema)
  resetSeed()
  return result
}

export { installFaker, seed, runFake, randexp } from './random'
export { install, installCustom } from './installation'
export { ZodSchemaFakerError } from './error'
export { ZodTypeFaker } from './zod-type-faker'
