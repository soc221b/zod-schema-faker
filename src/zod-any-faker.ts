import { z } from 'zod'
import { fake } from './fake'
import { getFaker } from './random'
import { ZodTypeFaker } from './zod-type-faker'

const schemas = (() => {
  const simpleSchemas = [
    z.string(),
    z.number(),
    z.boolean(),
    z.null(),
  ]
  const arraySchemas = simpleSchemas.map(schema => z.array(schema))
  const objectSchemas = simpleSchemas.map(schema =>
    z.object({
      foo: schema,
    }),
  )

  return [...simpleSchemas, ...arraySchemas, ...objectSchemas]
})()

export class ZodAnyFaker extends ZodTypeFaker<z.ZodAny> {
  fake(): z.infer<z.ZodAny> {
    const randomSchema = getFaker().helpers.arrayElement(schemas)
    return fake(randomSchema)
  }
}
