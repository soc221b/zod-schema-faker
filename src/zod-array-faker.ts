import { z } from 'zod'
import { fake } from './fake'
import { runFake } from './random'
import { ZodTypeFaker } from './zod-type-faker'

export class ZodArrayFaker<T extends z.ZodArray<any, any>> extends ZodTypeFaker<T> {
  fake(): z.infer<T> {
    const min = this.schema._def.exactLength?.value ?? this.schema._def.minLength?.value ?? 0
    const max =
      this.schema._def.exactLength?.value ??
      this.schema._def.maxLength?.value ??
      runFake(faker => faker.number.int({ min, max: min + 2 }))
    return runFake(faker => faker.helpers.multiple(() => fake(this.schema._def.type), { count: { min, max } }))
  }
}
