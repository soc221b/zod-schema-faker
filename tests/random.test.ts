import { expect, test } from 'vitest'
import { z } from 'zod'
import { install, fake, seed, runFake, randexp } from '../src'

test('runFake', () => {
  const data = runFake(faker => faker.datatype.boolean())
  expect(data).toBeTypeOf('boolean')
})

test('randexp', () => {
  const regex = /^foo|bar$/
  const data = randexp(regex)
  expect(data).toBeTypeOf('string')
  expect(data).toMatch(regex)
})

test('seed', () => {
  install()
  const schema = z.object({
    foo: z.number(),
    bar: z.number(),
    date: z.date(),
    string: z.object({
      date: z.string().date(),
      datetime: z.string().datetime(),
      time: z.string().time(),
    }),
  })

  seed(3)
  const data1 = fake(schema)
  seed(3)
  const data2 = fake(schema)
  expect(data1).toMatchSnapshot()
  expect(data1).toEqual(data2)

  const data3 = fake(schema)
  expect(data1).not.toEqual(data3)
})

test('runFake can be used with sync functions', () => {
  expect(() => runFake(() => {})).not.toThrow()
})

test('runFake can not be used with async functions', () => {
  // @ts-expect-error
  expect(() => runFake(async () => {})).toThrow()
})
