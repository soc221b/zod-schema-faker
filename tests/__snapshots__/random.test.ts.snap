// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`seed 1`] = `
{
  "bar": 915093660424243,
  "date": 2024-07-21T00:59:22.656Z,
  "foo": 915093660424243,
  "string": {
    "date": "5507-12-25",
    "datetime": "5507-12-25T00:42:12.465Z",
    "time": "00:42:12",
  },
}
`;
